File: Packages/Yage/Sources/Yage/Capabilities/WallCrawler.swift
import SwiftUI

public class WallCrawler: Capability {
    public override func install(on subject: Entity) {
        super.install(on: subject)
        subject.capability(for: BounceOnLateralCollisions.self)?.isEnabled = false
        subject.capability(for: FlipHorizontallyWhenGoingLeft.self)?.isEnabled = false
    }

    override public func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        guard case .move = subject?.state else { return }
        guard let direction = subject?.direction else { return }

        let isGoingUp = direction.dy < -0.0001
        let isGoingRight = direction.dx > 0.0001
        let isGoingDown = direction.dy > 0.0001
        let isGoingLeft = direction.dx < -0.0001

        if isGoingUp && touchesScreenTop() {
            crawlAlongTopBound()
            return
        }
        if isGoingRight && touchesScreenRight() {
            crawlUpRightBound()
            return
        }
        if isGoingDown && touchesScreenBottom() {
            crawlAlongBottomBound()
            return
        }
        if isGoingLeft && touchesScreenLeft() {
            crawlDownLeftBound()
            return
        }
    }

    private func crawlAlongTopBound() {
        guard let subject else { return }
        subject.direction = .init(dx: -1, dy: 0)
        subject.frame.origin = CGPoint(x: subject.frame.origin.x, y: 0)
        subject.rotation?.isFlippedHorizontally = true
        subject.rotation?.isFlippedVertically = true
        subject.rotation?.zAngle = 0
    }

    private func crawlUpRightBound() {
        guard let subject else { return }
        subject.direction = .init(dx: 0, dy: -1)
        subject.frame.origin = CGPoint(
            x: subject.worldBounds.width - subject.frame.width,
            y: subject.frame.origin.y
        )
        subject.rotation?.isFlippedHorizontally = false
        subject.rotation?.isFlippedVertically = false
        subject.rotation?.zAngle = .pi * 0.5
    }

    private func crawlAlongBottomBound() {
        guard let subject else { return }
        subject.direction = .init(dx: 1, dy: 0)
        subject.frame.origin = CGPoint(
            x: subject.frame.origin.x,
            y: subject.worldBounds.height - subject.frame.height
        )
        subject.rotation?.isFlippedHorizontally = false
        subject.rotation?.isFlippedVertically = false
        subject.rotation?.zAngle = 0
    }

    private func crawlDownLeftBound() {
        guard let subject else { return }
        subject.direction = .init(dx: 0, dy: 1)
        subject.frame.origin = CGPoint(x: 0, y: subject.frame.origin.y)
        subject.rotation?.isFlippedHorizontally = false
        subject.rotation?.isFlippedVertically = false
        subject.rotation?.zAngle = .pi * 1.5
    }

    private func touchesScreenTop() -> Bool {
        guard let subject else { return false }
        return subject.frame.minY <= 0
    }

    private func touchesScreenRight() -> Bool {
        guard let subject else { return false }
        let bound = subject.worldBounds.width
        return subject.frame.maxX >= bound
    }

    private func touchesScreenBottom() -> Bool {
        guard let subject else { return false }
        let bound = subject.worldBounds.height
        return subject.frame.maxY >= bound
    }

    private func touchesScreenLeft() -> Bool {
        guard let subject else { return false }
        return subject.frame.minX <= 0
    }
}

File: Sources/windows/yage/capabilities/wall_crawler.py
import math
from yage.models.capability import Capability
from yage.models.entity_state import EntityState
from yage.utils.geometry import Point, Vector


class WallCrawler(Capability):
    def do_update(self, collisions, time):
        if self.subject.state != EntityState.MOVE:
            return

        direction = self.subject.direction

        if self._is_going_up(direction) and self.touches_screen_top():
            self._crawl_along_top_bound()
            return
        if self._is_going_right(direction) and self.touches_screen_right():
            self._crawl_up_right_bound()
            return
        if self._is_going_down(direction) and self.touches_screen_bottom():
            self._crawl_along_bottom_bound()
            return
        if self._is_going_left(direction) and self.touches_screen_left():
            self._crawl_down_left_bound()
            return

    def _is_going_up(self, direction):
        return direction.dy < -0.0001

    def _is_going_right(self, direction):
        return direction.dx > 0.0001

    def _is_going_down(self, direction):
        return direction.dy > 0.0001

    def _is_going_left(self, direction):
        return direction.dx < -0.0001

    def _crawl_along_top_bound(self):
        direction = Vector(-1, 0)
        self.subject.direction = direction
        self.subject.frame.origin = Point(self.subject.frame.origin.x, 0)
        rotation = self.subject.rotation
        if rotation:
            rotation.is_flipped_horizontally = True
            rotation.is_flipped_vertically = True
            rotation.z_angle = 0

    def _crawl_up_right_bound(self):
        direction = Vector(0, -1)
        self.subject.direction = direction
        self.subject.frame.origin = Point(
            self.subject.world_bounds.width - self.subject.frame.width,
            self.subject.frame.origin.y
        )
        rotation = self.subject.rotation
        if rotation:
            rotation.is_flipped_horizontally = False
            rotation.is_flipped_vertically = False
            rotation.z_angle = math.pi * 0.5

    def _crawl_along_bottom_bound(self):
        direction = Vector(1, 0)
        self.subject.direction = direction
        self.subject.frame.origin = Point(
            self.subject.frame.origin.x,
            self.subject.world_bounds.height - self.subject.frame.height
        )
        rotation = self.subject.rotation
        if rotation:
            rotation.is_flipped_horizontally = False
            rotation.is_flipped_vertically = False
            rotation.z_angle = 0

    def _crawl_down_left_bound(self):
        direction = Vector(0, 1)
        self.subject.direction = direction
        self.subject.frame.origin = Point(0, self.subject.frame.origin.y)
        rotation = self.subject.rotation
        if rotation:
            rotation.is_flipped_horizontally = False
            rotation.is_flipped_vertically = False
            rotation.z_angle = math.pi * 1.5

    def touches_screen_top(self):
        return self.subject.frame.min_y <= 0

    def touches_screen_right(self):
        bound = self.subject.world_bounds.width
        return self.subject.frame.max_x >= bound

    def touches_screen_bottom(self):
        bound = self.subject.world_bounds.height
        return self.subject.frame.max_y >= bound

    def touches_screen_left(self):
        return self.subject.frame.min_x <= 0
